File: 11/solve.hs
module Main (main) where

import Control.Monad.State
import Data.Map qualified as M

rule :: Int -> [Int]
rule 0 = [1]
rule n =
  if even nd
    then let (d, m) = divMod n (10 ^ (nd `div` 2)) in [d, m]
    else [n * 2024]
  where
    nd = nDigits n

nDigits :: Int -> Int
nDigits = length . takeWhile (> 0) . iterate (`div` 10)

type Cached a = State (M.Map (Int, Int) Int) a

lengthAfterNSteps :: Int -> Int -> Cached Int
lengthAfterNSteps 0 _ = pure 1
lengthAfterNSteps steps num = do
  cache <- get
  case M.lookup (steps, num) cache of
    Just len -> pure len
    Nothing -> do
      res <- sum <$> mapM (lengthAfterNSteps (steps - 1)) (rule num)
      modify (M.insert (steps, num) res)
      pure res

evalLength :: Int -> [Int] -> Int
evalLength steps nums =
  flip evalState M.empty $ sum <$> mapM (lengthAfterNSteps steps) nums

parseInput :: String -> [Int]
parseInput = map read . words

part1 :: [Int] -> Int
part1 = evalLength 25

part2 :: [Int] -> Int
part2 = evalLength 75

main :: IO ()
main = do
  input <- parseInput <$> readFile "input.txt"
  print $ part1 input
  print $ part2 input

File: 11/solve.py
import functools
from typing import Iterable


def rule(n: int) -> tuple[int, ...]:
    if n == 0:
        return (1,)
    d, m = divmod(n_digits(n), 2)
    if m == 0:
        return divmod(n, 10**d)
    else:
        return (n * 2024,)


def n_digits(a: int) -> int:
    n = 0
    while a > 0:
        n += 1
        a //= 10
    return n


@functools.cache
def length_after_n_steps(steps: int, num: int) -> int:
    if steps == 0:
        return 1
    return sum_length_after_n_steps(steps - 1, rule(num))


def sum_length_after_n_steps(steps: int, nums: Iterable[int]) -> int:
    return sum(length_after_n_steps(steps, num) for num in nums)


def parse_input(inp: str) -> list[int]:
    return list(map(int, inp.split()))


def part1(inp: list[int]) -> int:
    return sum_length_after_n_steps(25, inp)


def part2(inp: list[int]) -> int:
    return sum_length_after_n_steps(75, inp)


def main() -> None:
    with open("input.txt") as f:
        inp = parse_input(f.read())

    print(part1(inp))
    print(part2(inp))


if __name__ == "__main__":
    main()
